### PR TITLE
Fix all server side admin actions, improve test coverage 

### PR DESCRIPTION
**Pull request by goldenapples**
With this pull request, all admin functionality related to adding, editing, and reordering should be usable and the PHP portion of it covered by tests. I'm working on setting up a test suite for `coauthors.js`, and also separating it into classes, refactoring, and improving the look and feel of the UI. I will put that into a different pull request, because I think it probably needs to be evaluated separately.

### CODE REVIEW

File: includes/admin-edit-ui.php [hunk 1]
```diff
@@ -172,11 +176,11 @@ function template_coauthor_sortable( $coauthor, $author_role = null ) {
 					<span class="author-name"><?php echo $coauthor->display_name; ?></span>
 					<span class="author-email"><?php echo $coauthor->user_email; ?></span>
 				</span>
-				<?php $author_role = ( isset( $coauthor->author_role ) ) ?
+				<?php $author_role = ( ! empty( $coauthor->author_role ) ) ?
```
Comment (danielbachhuber) [inline includes/admin-edit-ui.php]: `'BYLINE'` should be translatable.

File: includes/admin-edit-ui.php [hunk 2]
```diff
@@ -204,7 +208,7 @@ function ajax_template_coauthor_sortable() {
 	check_ajax_referer( 'coauthor-select', '_ajax_coauthor_template_nonce' );
 	global $coauthors_plus;
 
-	$coauthor = $coauthors_plus->get_coauthor_by( 'id', intval( $_REQUEST['authorId'] ) );
+	$coauthor = $coauthors_plus->get_coauthor_by( 'user_nicename', $_REQUEST['authorId'] );
```
Comment (danielbachhuber) [inline includes/admin-edit-ui.php]: `$_REQUEST['authorId']` needs to be sanitized before use

File: includes/admin-edit-ui.php [hunk 3]
```diff
@@ -159,34 +163,38 @@ function template_coauthor_sortable( $coauthor, $author_role = null ) {
...
 				<span class="author-info sortable-flex-section">
-					<span class="author-name"><?php echo $coauthor->display_name; ?></span>
-					<span class="author-email"><?php echo $coauthor->user_email; ?></span>
+					<span class="author-name"><?php echo esc_html( $coauthor->display_name ); ?></span>
+					<span class="author-email"><?php echo esc_html( $coauthor->user_email ); ?></span>
 				</span>
-				<?php $author_role = ( isset( $coauthor->author_role ) ) ?
-										$coauthor->author_role : 'BYLINE'; ?>
+				<?php $author_role = ( ! empty( $coauthor->author_role ) ) ?
+										$coauthor->author_role : esc_html__( 'BYLINE', 'co-authors-plus-roles' ); ?>
```
Comment (danielbachhuber) [inline includes/admin-edit-ui.php]: Can we also uppercase this with CSS? Uppercase in code always makes me feel dirty

File: includes/admin-edit-ui.php [hunk 4]
```diff
@@ -204,8 +212,8 @@ function ajax_template_coauthor_sortable() {
 	check_ajax_referer( 'coauthor-select', '_ajax_coauthor_template_nonce' );
 	global $coauthors_plus;
 
-	$coauthor = $coauthors_plus->get_coauthor_by( 'id', intval( $_REQUEST['authorId'] ) );
-	$role = get_author_role( $_REQUEST['authorRole'] );
+	$coauthor = $coauthors_plus->get_coauthor_by( 'user_nicename', sanitize_text_field( $_REQUEST['authorId'] ) );
```
Comment (danielbachhuber) [inline includes/admin-edit-ui.php]: If this is no longer the `id`, shouldn't we rename the field?

File: includes/author-roles-posts-relationships.php [hunk 5]
```diff
@@ -48,34 +49,60 @@ function set_author_on_post( $post_id, $author, $author_role = false ) {
...
-	);
+	foreach ( get_post_meta( $post_id ) as $key => $values ) {
+		if ( strpos( $key, 'cap-' ) === 0 && in_array( $author->user_nicename, $values ) ) {
+			delete_post_meta( $post_id, $key, $author->user_nicename );
+		}
+	}
 
 	if ( ! is_object( $author_role ) ) {
-		$author_role = get_author_role( $author_role );
+		$author_role = get_author_role( sanitize_text_field( $author_role ) );
```
Comment (danielbachhuber) [inline includes/author-roles-posts-relationships.php]: I think sanitizing here might be overzealous

File: includes/author-roles-posts-relationships.php [hunk 6]
```diff
@@ -48,34 +49,60 @@ function set_author_on_post( $post_id, $author, $author_role = false ) {
...
 	}
 
 	if ( ! $author_role ) {
 		return false;
 	}
 
-	update_post_meta( $post_id, 'cap-' . $author_role->slug, $author->user_nicename );
+	add_post_meta( $post_id,
+		sanitize_key( 'cap-' . $author_role->slug ),
+		sanitize_user( $author->user_nicename )
```
Comment (danielbachhuber) [inline includes/author-roles-posts-relationships.php]: Same as above

File: includes/query.php [hunk 7]
```diff
@@ -74,30 +75,31 @@ function get_coauthors( $post_id = 0, $args = array() ) {
 		if ( $args['author_role'] === 'any' || ! $args['author_role'] ) {
 			$author_roles = wp_list_pluck( get_author_roles(), 'slug' );
 		} else if ( is_string( $args['author_role'] ) ) {
-			$author_roles = array( $args['author_role'] );
+			$author_roles = array( sanitize_text_field( $args['author_role'] ) );
 		} else {
-			$author_roles = $args['author_role'];
+			$author_roles = array_map( 'sanitize_text_field', $args['author_role'] );
```
Comment (danielbachhuber) [inline includes/query.php]: Same as above — not sure we need sanitizing here.

File: includes/query.php [hunk 8]
```diff
@@ -117,14 +121,16 @@ function( $term ) { return 'cap-' . $term; },
 					}
 
 					$post_author = $coauthors_plus->get_coauthor_by( 'user_nicename', $coauthor_slug );
-					$post_author->author_role = 'byline';
+					$post_author->author_role = __( 'BYLINE', 'co-authors-plus-roles' );
```
Comment (danielbachhuber) [inline includes/query.php]: `esc_html__()` and no caps

File: includes/admin-edit-ui.php (hunk 3 above)
Comment (goldenapples) [inline includes/admin-edit-ui.php]: Sure. I was trying to set it apart to indicate that it's not treated like other values in that field. The other values available for that field are all keys from the `$_wp_author_roles` object, but 'BYLINE' is not, and is handled differently. (It's saved as an empty string, for compatibility with CAP.) 

Adding liberal comments wherever that value is referenced would achieve the same purpose with less mess.

File: includes/admin-edit-ui.php (hunk 4 above)
Comment (goldenapples) [inline includes/admin-edit-ui.php]: Agreed. Now that I have enough tests in place to safely do that, I should be fine renaming it.

**Comment by goldenapples** [comment]
Updated to respond to code review comments. Removed superfluous sanitization and renamed author identifier variables for clarity (s/id/nicename).

**Comment by danielbachhuber** [comment]
:+1:

**Pull request merged**